components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+                },
+                (State::AfterTypeName, _) => Err(()),
+
+                (State::AfterType { .. }, &Token::WhiteSpace(_)) => {
+                    // Ignore whitespace.
+                    Ok(State::AfterType { after_whitespace: true })
+                },
+                (State::AfterType { after_whitespace: false }, &Token::Delim('+')) => {
+                    // We should be following some type.
+                    // We should panic if we're not, because we should only get
+                    // here from Start -> AfterOpenAngle -> AfterTypeName (in
+                    // the case of something like '<length>') or
+                    // Start (in the case of something like 'my-ident'), both
+                    // of which should have pushed a type.
+                    if let Some(Syntax::Disjunction(ref mut ts)) = *syntax {
+                        let term = &mut ts.last_mut().unwrap();
+                        if term.typ == Type::TransformList {
+                            // <transform-list>+ is specifically disallowed.
+                            return Err(())
+                        }
+                        term.list = true
+                    } else { unreachable!() }
+                    Ok(State::AfterPlus)
+                },
+                (State::AfterType { .. }, &Token::Delim('|')) => {
+                    // Some other term in the disjunction should follow.
+                    Ok(State::Start { after_bar: true })
+                },
+                (State::AfterType { .. }, _) => Err(()),
+
+                (State::AfterAsterisk, &Token::WhiteSpace(_)) => Ok(State::AfterAsterisk),
+                (State::AfterAsterisk, _) => Err(()),
+                (State::AfterPlus, &Token::WhiteSpace(_)) => Ok(State::AfterPlus),
+                (State::AfterPlus, &Token::Delim('|')) => {
+                    // Some other term in the disjunction should follow.
+                    Ok(State::Start { after_bar: true })
+                }
+                (State::AfterPlus, _) => Err(()),
+            }
+        }
+
+        // Loop over all of the tokens in the syntax string.
+        loop {
+            match parser.next_including_whitespace_and_comments() {
+                Err(BasicParseError::EndOfInput) => {
+                    match state {
+                        State::Start { after_bar: false } |
+                        State::AfterType { .. } |
+                        State::AfterAsterisk |
+                        State::AfterPlus => break,
+
+                        // We shouldn't reach EOF in the middle of something.
+                        State::Start { after_bar: true } |
+                        State::AfterOpenAngle |
+                        State::AfterTypeName => return Err(())
+                    }
+                },
+                Err(_) => return Err(()),
+                Ok(token) => {
+                    match handle_token(&mut syntax, state, token) {
```

Comment:
nit: You can write this like:

```rust
state = handle_token(&mut syntax, state, token)?;
```

Not sure how you feel about the extra magic.